### PR TITLE
DOC: improve `issubdtype` and scalar type docs

### PR DESCRIPTION
**Pull request by eric-wieser**
Partially in response to #17325, cc @rgommers.

https://15776-908607-gh.circle-artifacts.com/0/doc/build/html/reference/generated/numpy.issubdtype.html

https://15776-908607-gh.circle-artifacts.com/0/doc/build/html/reference/arrays.scalars.html

### CODE REVIEW

File: doc/source/reference/arrays.scalars.rst [hunk 1]
```diff
@@ -65,19 +65,20 @@ Some of the scalar types are essentially equivalent to fundamental
...
-:class:`unicode_`     :class:`UnicodeType`
-====================  ================================
+====================  ===================  =============
+Array scalar type     Related Python type  Inherits?
+====================  ===================  =============
+:class:`int_`         :class:`int`         Python 2 only
+:class:`float_`       :class:`float`       yes
+:class:`complex_`     :class:`complex`     yes
+:class:`bytes_`       :class:`bytes`       yes
+:class:`unicode_`     :class:`unicode`     yes
```
Comment (BvB93) [inline doc/source/reference/arrays.scalars.rst:75]:
```suggestion
:class:`str_`         :class:`str`         yes
```
Considering `unicode` was removed in python 3 might it not be better to link to `str` instead?

Comment (eric-wieser) [reply doc/source/reference/arrays.scalars.rst:75]: Good catch

File: doc/source/reference/arrays.scalars.rst [hunk 2]
```diff
@@ -65,19 +65,20 @@ Some of the scalar types are essentially equivalent to fundamental
...
-====================  ================================
-:class:`int_`         :class:`IntType` (Python 2 only)
-:class:`float_`       :class:`FloatType`
-:class:`complex_`     :class:`ComplexType`
-:class:`bytes_`       :class:`BytesType`
-:class:`unicode_`     :class:`UnicodeType`
-====================  ================================
+====================  ===================  =============
+Array scalar type     Related Python type  Inherits?
+====================  ===================  =============
```
Comment (BvB93) [inline doc/source/reference/arrays.scalars.rst:70]: `np.datetime64` and `np.timedelta64` are still missing.
`np.void` _might_ also be worth mentioning, even though it's strictly speaking not a scalar.

Comment (eric-wieser) [reply doc/source/reference/arrays.scalars.rst:70]: This table is not exhaustive, it only covers things with related builtin types. Void doesn't have a corresponding builtin type, but I suppose `np.datetime64` and `np.timedelta64` do: added.

**Comment by eric-wieser** [comment]
Not sure why github is linking to the wrong CircleCI build, the real one is https://app.circleci.com/pipelines/github/numpy/numpy/3679/workflows/bbd2257f-2d41-4cb9-882a-48511446daa7/jobs/15768

**Comment by seberg** [comment]
@larsoner a ping about the build artifact link. It seems it is sometimes broken (such as here), although it is possible that this only happens when circle CI gives an error.

**Comment by eric-wieser** [comment]
@seberg: In this case even the first-party CircleCI link is broken

File: numpy/core/numerictypes.py [hunk 3]
```diff
@@ -373,14 +375,44 @@ def issubdtype(arg1, arg2):
     See Also
     --------
     issubsctype, issubclass_
-    numpy.core.numerictypes : Overview of numpy type hierarchy.
+    arrays.scalars : Overview of numpy type hierarchy.
```
Comment (eric-wieser) [inline numpy/core/numerictypes.py:378]: This link still doesn't work - any ideas?

**Comment by seberg** [comment]
Oh, OK, then maybe that type of thing caused it to fail somewhere. It was just a single PR where I had noticed a mismatch, thought it was repeating now.

**Comment by larsoner** [comment]
There can be problems when the fork and upstream both have CircleCI enabled:

https://github.com/larsoner/circleci-artifacts-redirector-action/issues/7

But indeed these builds are just failing themselves

File: numpy/core/numerictypes.py (hunk 3 above)
Comment (seberg) [reply numpy/core/numerictypes.py:378]: Ping @rossbar, who may know.

Comment (rgommers) [reply numpy/core/numerictypes.py:378]: This links to code objects. `numpy.core.numerictypes` is a thing, `arrays.scalars` is not. You'll need `:ref:` I think.

File: numpy/core/numerictypes.py [hunk 4]
```diff
@@ -373,14 +375,44 @@ def issubdtype(arg1, arg2):
...
+    False
+
+    but both are subtypes of `floating`:
+
+    >>> np.issubdtype(np.float64, np.floating)
+    True
+    >>> np.issubdtype(np.float32, np.floating)
+    True
+
+    For convinience, dtype-like objects are allowed too:
```
Comment (rgommers) [inline numpy/core/numerictypes.py:410]: minor: typo here

**Comment by rgommers** [comment]
The improved `issubdtype` docstring with examples looks much better, thanks @eric-wieser

File: numpy/core/numerictypes.py (hunk 3 above)
Comment (eric-wieser) [reply numpy/core/numerictypes.py:378]: `numpy.core.numerictypes` is not a thing that exists in the generated documentation.

Comment (eric-wieser) [reply numpy/core/numerictypes.py]: Hopefully fixed, let's see...

**Comment by mattip** [comment]
This needs the fix in gh-17304 to pin pygments version

**Comment by eric-wieser** [comment]
Will rebase on master if this needs any other changes - otherwise I'd be inclined to merge as is rather than starting another CI run.

File: numpy/core/numerictypes.py (hunk 3 above)
Comment (eric-wieser) [reply numpy/core/numerictypes.py]: Yep, fixed

**Comment by seberg** [comment]
The artifact for the "See also" link seems correct and the rest also seems fine and is definitely an improvement, so I will just put this in (there are also no warnings beyond the gh-17304 one). Thanks.

**Pull request merged**